### PR TITLE
Allow context-specific CRT verification callbacks

### PR DESCRIPTION
**Pull request by hanno-arm**
__Summary:__ This PR adds a new SSL API `mbedtls_ssl_set_verify()` which is analogous to the existing `mbedtls_ssl_conf_verify()` but binds the configured CRT verification callback to the SSL context and not to the SSL configuration.

This is useful for example when the verification callback not only performs generic checks but also needs to remember parts of the peer's CRT chain in a connection-specific location.

__Note:__ I suggest we add a compile-time option guarding the entire CRT verification callback functionality (passing callbacks to `mbedtls_x509_crt_verify()`, `mbedtls_ssl_conf_verify()`, and `mbedtls_ssl_set_verify()`) in order to save code on constrained systems that don't any variant of the verification callback feature, but that can be discussed and implemented elsewhere.

### CODE REVIEW

**Comment by hanno-arm** [comment]
Travis passed, but result isn't communicated back to GitHub. Restarting.

**Comment by yanesca** [comment]
Travis has failed on DTLS proxy tests, which is a known issue.

File: tests/ssl-opt.sh [hunk 1]
```diff
@@ -1008,7 +1008,21 @@ run_test    "Unique IV in GCM" \
...
             -U "IV used"
 
-# Test for context-specific CRT verification callback
+# Tests for certificate verification callback
+run_test    "Configuration-specific CRT verification callback" \
+            "$P_SRV debug_level=3" \
+            "$P_CLI context_crt_cb=0 debug_level=3" \
+            0 \
+            -s "Protocol is TLSv1.2" \
+            -s "Ciphersuite is TLS-ECDHE-RSA-WITH-CHACHA20-POLY1305-SHA256" \
```
Comment (Patater) [inline tests/ssl-opt.sh]: Is this a bit fragile? If the list of default ciphersuites change, won't this also need changing?

Comment (hanno-arm) [reply tests/ssl-opt.sh]: Yes, that's a general problem we have with `ssl-opt.sh` - the test template is taken from the `Default` test, which has the same check. However, there's no reason not to remove this line in this test, to not introduce yet another test which isn't compatible with reduced configs.

File: programs/ssl/ssl_client2.c [hunk 2]
```diff
@@ -126,6 +127,16 @@ int main( void )
...
 
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+#define USAGE_CALLBACK \
+    "    context_crt_cb=%%d   This determines whether the CRT verification callback is bound\n" \
+    "                        to the SSL configuration of the SSL context.\n" \
+    "                        Possible values:\n"\
+    "                        - 0 (default): Use CRT callback bound to configuration\n" \
+    "                        - 1: Use CRT callback bound to SSL context\n"
+#else
+#define USAGE_CALLBACK ""
```
Comment (Patater) [inline programs/ssl/ssl_client2.c]: Is there a better name to use than "callback"? Perhaps `USAGE_CONTEXT_CRT_CB` to match the variable name and `DFL_CONTEXT_CRT_CB`.

File: tests/ssl-opt.sh (hunk 1 above)
Comment (mpg) [reply tests/ssl-opt.sh]: There are many problems with `ssl-opt.sh` but I don't think this is one of them. IMO it's rather a problem of insufficient editing after copy-paste, which can happen with any script and to anyone :)

Comment (hanno-arm) [reply tests/ssl-opt.sh]: @mpg I didn't mean to defend the copy-paste - it should be changed - but the `Default` test _does_ have the same problem, doesn't it? It hardcodes the expectation of ChaChaPoly as the chosen ciphersuite.

Comment (mpg) [reply tests/ssl-opt.sh]: Yes, the default test clearly requires a bunch more `requires_config`, as do most tests here, which is a big problem with `ssl-opt.sh` indeed.

However the way I see it, the hardcoding it intentional here: we want to test what our default behaviour is (hence the "Default" in the test name), and the alternative to hardcoding is to have the test system also compute what the selected algorithms should be based on `config.h`, but IMO that's fragile in another way: what if we make the same mistake in computing those in the test script and in the library? Hardcoding, with all the issues it has, is more robust in that respect.

But that's largely out of scope here, and probably a discussion we'll need to have some other time.

File: programs/ssl/ssl_client2.c [hunk 3]
```diff
@@ -326,6 +337,7 @@ int main( void )
     USAGE_TICKETS                                           \
     USAGE_MAX_FRAG_LEN                                      \
     USAGE_TRUNC_HMAC                                        \
+    USAGE_CONTEXT_CRT_CB                                          \
```
Comment (Patater) [inline programs/ssl/ssl_client2.c]: Style: Alignment is funny here

Comment (yanesca) [reply programs/ssl/ssl_client2.c]: I have amended the commit.

**Comment by szysas** [comment]
@hanno-arm is it possible to have a branch with this feature rebased on top of 2.16.0?

**Comment by yanesca** [comment]
@szysas I hope that we can resolve the issue with 2.17 pretty soon, but if that is not the case, then I am happy to rebase it as a temporary solution.

(Hanno has handed this PR over to me last week.)

**Comment by szysas** [comment]
At the moment, rebasing on top of 2.17 is fine.

**Comment by yanesca** [comment]
@Patater , @dgreen-arm Do you have a review in progress? Do you mind if I rebase it?

**Comment by szysas** [comment]
@yanesca maybe lets use latest (2.17) as base

**Pull request merged**